File: contracts/emotion_state.py
import time
from dataclasses import dataclass, field, asdict
from typing import Dict, Any

@dataclass
class EmotionState:
    """
    Standardized EmotionState Data Contract (v1.0.0).
    Contains primary emotion, secondary emotions, intensity values,
    valence, arousal, dominance, mood baseline, emotional momentum, and decay rates.
    """
    version: str = "1.0.0"
    timestamp: float = field(default_factory=time.time)
    primary_emotion: str = "neutral"
    secondary_emotions: Dict[str, float] = field(default_factory=dict)
    intensity_values: Dict[str, float] = field(default_factory=dict)
    valence: float = 0.0      # [-1.0 to +1.0]
    arousal: float = 0.5      # [ 0.0 to  1.0]
    dominance: float = 0.5    # [ 0.0 to  1.0]
    mood_baseline: str = "calmness"
    emotional_momentum: float = 1.0
    decay_rate: float = 7200.0

    def to_dict(self) -> Dict[str, Any]:
        return asdict(self)

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> "EmotionState":
        if not data:
            return cls()
        known = {k: v for k, v in data.items() if k in cls.__dataclass_fields__}
        return cls(**known)

File: Mate-Engine/Assets/MATE ENGINE - Scripts/Contracts/EmotionState.cs
using System;
using System.Collections.Generic;
using UnityEngine;

namespace Vivy.Contracts
{
    /// <summary>
    /// Standardized EmotionState Data Contract (v1.0.0).
    /// Mirror of Python EmotionState dataclass.
    /// </summary>
    [Serializable]
    public class EmotionState
    {
        public string version = "1.0.0";
        public double timestamp = 0;
        public string primary_emotion = "neutral";
        public SerializableDictionary<string, float> secondary_emotions = new SerializableDictionary<string, float>();
        public SerializableDictionary<string, float> intensity_values = new SerializableDictionary<string, float>();
        public float valence = 0.0f;
        public float arousal = 0.5f;
        public float dominance = 0.5f;
        public string mood_baseline = "calmness";
        public float emotional_momentum = 1.0f;
        public float decay_rate = 7200.0f;

        public static EmotionState FromJson(string json)
        {
            if (string.IsNullOrEmpty(json)) return new EmotionState();
            try
            {
                return JsonUtility.FromJson<EmotionState>(json);
            }
            catch (Exception ex)
            {
                Debug.LogWarning($"[Contracts] Failed to parse EmotionState JSON: {ex.Message}");
                return new EmotionState();
            }
        }

        public string ToJson()
        {
            return JsonUtility.ToJson(this);
        }
    }

    [Serializable]
    public class SerializableDictionary<TKey, TValue> : ISerializationCallbackReceiver
    {
        [SerializeField] private List<TKey> keys = new List<TKey>();
        [SerializeField] private List<TValue> values = new List<TValue>();

        private Dictionary<TKey, TValue> dictionary = new Dictionary<TKey, TValue>();

        public Dictionary<TKey, TValue> Dict => dictionary;

        public void OnBeforeSerialize()
        {
            keys.Clear();
            values.Clear();
            foreach (var pair in dictionary)
            {
                keys.Add(pair.Key);
                values.Add(pair.Value);
            }
        }

        public void OnAfterDeserialize()
        {
            dictionary.Clear();
            for (int i = 0; i < Math.Min(keys.Count, values.Count); i++)
            {
                dictionary[keys[i]] = values[i];
            }
        }
    }
}
